lxd/acme/acme.go

```diff
@@ -0,0 +1,127 @@
+package acme
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/certificate"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// UpdateACMECertificate updates the ACME certificate.
+func UpdateACMECertificate(s *state.State, provider HTTP01Provider, clustered bool, domain string, email string, caURL string) (*certificate.Resource, error) {
+	// Load the certificate.
+	certInfo, err := util.LoadCert(s.OS.VarDir)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load certificate and key file: %w", err)
+	}
+
+	cert, err := x509.ParseCertificate(certInfo.KeyPair().Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+	}
+
+	issueNewCertificate := true
+
+	// If certificate domain matches the configured domain, renew the cert.
+	if shared.StringInSlice(domain, cert.DNSNames) {
+		issueNewCertificate = false
+	}
+
+	// Check the certificate's expiration date, and renew it if it's only valid for less than 30 days.
+	if !issueNewCertificate && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30)) {
+		return nil, nil
+	}
+
+	// Generate new private key for user. This key needs to be different from the server's private key.
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("Failed generating private key for user account: %w", err)
+	}
+
+	user := user{
+		Email: email,
+		Key:   privateKey,
+	}
+
+	config := lego.NewConfig(&user)
+
+	if caURL != "" {
+		config.CADirURL = caURL
+	} else {
+		// Default URL for Let's Encrypt
+		config.CADirURL = "https://acme-v02.api.letsencrypt.org/directory"
+	}
+
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	client, err := lego.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create new client: %w", err)
+	}
+
+	err = client.Challenge.SetHTTP01Provider(provider)
+	if err != nil {
+		return nil, fmt.Errorf("Failed setting HTTP-01 provider: %w", err)
+	}
+
+	var reg *registration.Resource
+
+	// Registration might fail randomly (as seen in manual tests), so retry in that case.
+	// The number 5 is chosen here because Let's Encrypt has a limit of 5 failures per account,
+	// per hostname, per hour.
+	for i := 0; i < 5; i++ {
+		reg, err = client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
```

Comment:
`TermsOfServiceAgreed` should probably be a config setting that the user has to set to say they've agreed with the terms. Otherwise we are implicitly agreeing on their behalf.

Reply:
Do we have to register every time we issue a new cert, or should we record that the registration has been successful the first time?

Reply:
> TermsOfServiceAgreed should probably be a config setting that the user has to set to say they've agreed with the terms. Otherwise we are implicitly agreeing on their behalf.

If this is important, we can surely do that.

> Do we have to register every time we issue a new cert, or should we record that the registration has been successful the first time?

I discussed this with @stgraber, and we agreed that we won't store the account information as an account isn't really beneficial for us. Therefore, we'll register every time.